crates/oxc_linter/src/utils/react.rs

```diff
@@ -351,3 +352,146 @@ pub fn is_state_member_expression(expression: &StaticMemberExpression<'_>) -> bo
+            if function_contains_jsx(func) { Some(InnermostFunction::Function(func)) } else { None }
+        }
+        Expression::ArrowFunctionExpression(arrow_func) => {
+            // Check if this arrow function contains JSX
+            if expression_contains_jsx(expr) {
+                Some(InnermostFunction::ArrowFunction(arrow_func))
+            } else {
+                // Check if this arrow function returns another function that contains JSX
+                if arrow_func.expression {
+                    // Expression-bodied arrow function: () => () => <div />
+                    if arrow_func.body.statements.len() == 1
+                        && let Statement::ExpressionStatement(expr_stmt) =
+                            &arrow_func.body.statements[0]
+                    {
+                        return find_innermost_function_with_jsx(&expr_stmt.expression, ctx);
+                    }
+                } else {
+                    // Block-bodied arrow function: () => { return () => <div /> }
+                    for stmt in &arrow_func.body.statements {
+                        if let Statement::ReturnStatement(ret_stmt) = stmt
+                            && let Some(expr) = &ret_stmt.argument
+                        {
+                            return find_innermost_function_with_jsx(expr, ctx);
+                        }
+                    }
+                }
+                None
+            }
+        }
+        _ => None,
+    }
+}
+
+/// Checks if an expression contains JSX elements
+pub fn contains_jsx(expr: &Expression) -> bool {
+    match expr {
+        Expression::JSXElement(_) | Expression::JSXFragment(_) => true,
+        Expression::CallExpression(call) => {
+            if crate::utils::is_create_element_call(call) {
+                return true;
+            }
+            call.arguments.iter().any(|arg| arg.as_expression().is_some_and(contains_jsx))
+        }
+        Expression::ParenthesizedExpression(inner) => contains_jsx(&inner.expression),
+        Expression::StaticMemberExpression(member) => contains_jsx(&member.object),
+        Expression::ConditionalExpression(cond) => {
+            contains_jsx(&cond.consequent) || contains_jsx(&cond.alternate)
+        }
+        Expression::LogicalExpression(logical) => {
+            contains_jsx(&logical.left) || contains_jsx(&logical.right)
+        }
+        Expression::SequenceExpression(seq) => seq.expressions.iter().any(contains_jsx),
+        _ => false,
+    }
+}
+
+/// Checks if a function contains JSX in its return statements
+pub fn function_contains_jsx(func: &Function) -> bool {
+    if let Some(body) = &func.body {
+        for stmt in &body.statements {
+            if let Statement::ReturnStatement(ret_stmt) = stmt
+                && let Some(expr) = &ret_stmt.argument
+                && contains_jsx(expr)
+            {
+                return true;
+            }
+        }
+    }
+    false
+}
```

Comment:
this function won't work properly as soon as the return is in a block statment.
e.g.
```
function Foo() {
    if (true) {
        return <div></div>
    }
}